pkg/apis/core/types.go

```diff
@@ -2124,6 +2134,11 @@ type Lifecycle struct {
 	// or until the termination grace period is reached.
 	// +optional
 	PreStop *Handler
+	// Type describes the startup/shutdown behaviour of the container.
+	// One of "Standard", "Sidecar".
+	// Defaults to "Standard".
```

Comment:
You say it default to Standard, but that is not implemented anywhere - there should be a corresponding change to pkg/apis/core/v1/defaults.go that explicitly sets the value, right?

Reply:
I wasn't entirely sure about the convention on this. How much should we be defaulting for a new alpha field? Currently `Lifecycle` is not a default field on a container so should we be adding `Lifecycle` + `Lifecycle.Type` to all containers and setting that to `Standard`? Or should we only default `Lifecycle.Type` if `Lifecycle != nil`?

Reply:
I also seem to recall some issues with defaulting PodSpec values as it would cause rollouts on a version upgrade, such as in https://github.com/kubernetes/kubernetes/pull/78885 and https://github.com/kubernetes/kubernetes/pull/69988

Reply:
@liggitt since you were active on #78885

What is the correct course for default fields that end up underneath pods?  It seems that the pattern from #69988 is to set if for Pod but not all PodSpec. (frankly, the type-based defaulting was/is a bad idea, we should fix that).

Reply:
Defaults in podspecs are particularly twisty. The safest thing to do is to avoid any defaulting that would change values in existing pods. For example, a default value for a field like this would could change content for existing pods. A default value for a field inside a new stanza (e.g. a new volume source type) would not change existing pods, because no existing pods would have that stanza.

The three areas podspec defaults have bitten us are:

1. New fields in podspec defaulted inside a deployment/replicaset/etc making workload controllers trigger new rollouts on API server upgrade. From what I understand, this is due to fragile computation by workload controllers of when the embedded pod spec has "changed" (e.g. hashing the pod spec, which is confused when a new field with a default value appears). This test catches additions like that:

https://github.com/kubernetes/kubernetes/blob/66262e351dad61cd70bd1591c7e69a3e5d218a28/pkg/apis/core/v1/defaults_test.go#L43-L52

2. New fields in podspec defaulted inside a pod (especially inside a container) making nodes restart pods on API server upgrade, due to the way the node detects whether the container spec it was asked to run has changed. This test catches additions like that:

https://github.com/kubernetes/kubernetes/blob/66262e351dad61cd70bd1591c7e69a3e5d218a28/pkg/apis/core/v1/defaults_test.go#L164-L172

3. New fields appearing anywhere within the podspec schema on API server upgrade (whether the field had a value or not) used to make nodes restart pods because they used to hash the go schema; that was fixed in https://github.com/kubernetes/kubernetes/pull/57741 in 1.16

Reply:
So in this scenario, if we were to default `Lifecycle.Type` that would cause changes to existing Pods as we would be changing the values in `Lifecycle`, hence causing restarts/rollouts. Therefore we shouldn't be doing any defaulting here to avoid that from happening. 
I can change the comments etc, so that there is no reference to `Standard` being the default, just that it can be `Sidecar` or `Standard`. There are no assumptions in the Kubelet code that non-sidecars will have `Standard` set on them, so it isn't an issue from that perspective.

Reply:
@thockin, any thoughts on this? I believe this is the last point in your review so once we've decided how to proceed with this, hopefully we'll be done :)

Reply:
I hate the limitation.  Hashing all of podspec is a terrible idea.  That said, it's what happens.

If we do not actually set the value, can we can assert that unset means "normal".  That makes it awkward if we ever want to change the default in a future API version, but that seems quite unlikely.

If that's the path, and it sounds like that is what @liggitt is saying, then we should document that and remove the "Standard" value entirely.

Reply:
ok thanks that clear things up, so I shall remove "Standard" as it's somewhat redundant, and document that unset means you're following the "normal" lifecycle as opposed to the sidecar one, the kubelet code already makes this assumption so this will be a simple change to make. Thanks for the quick response :)

Reply:
Ok, I've updated this to remove the "standard" lifecycle type, if you're now happy with this PR, we can close this and move over to the implementation PR https://github.com/kubernetes/kubernetes/pull/80744